File: telemetry/aitf/sdk/go/instrumentation/drift_detection.go
// Package instrumentation provides AITF instrumentation for AI operations.
package instrumentation

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"

	"github.com/girdav01/AITF/sdk/go/semconv"
)

const driftTracerName = "instrumentation.drift_detection"

// DriftDetectionInstrumentor traces model drift detection, baseline management,
// investigation, and remediation operations. Aligned with CoSAI's identification
// of model drift as a top-level AI incident category.
type DriftDetectionInstrumentor struct {
	tracer trace.Tracer
}

// NewDriftDetectionInstrumentor creates a new drift detection instrumentor.
func NewDriftDetectionInstrumentor(tp trace.TracerProvider) *DriftDetectionInstrumentor {
	if tp == nil {
		tp = otel.GetTracerProvider()
	}
	return &DriftDetectionInstrumentor{
		tracer: tp.Tracer(driftTracerName),
	}
}

// ── Detection ───────────────────────────────────────────────────────────────

// DetectConfig holds configuration for a drift detection span.
type DetectConfig struct {
	ModelID          string
	DriftType        string // e.g. "data_distribution", "concept", "performance"
	DetectionMethod  string // e.g. "psi", "ks_test", "js_divergence"
	ReferenceDataset string
	ReferencePeriod  string
	Threshold        *float64
}

// DriftDetection provides methods for enriching a drift detection span.
type DriftDetection struct {
	span trace.Span
}

// TraceDetect starts a span for a drift detection analysis. The caller must call
// End() when done.
func (d *DriftDetectionInstrumentor) TraceDetect(ctx context.Context, cfg DetectConfig) (context.Context, *DriftDetection) {
	spanName := fmt.Sprintf("drift.detect %s %s", cfg.DriftType, cfg.ModelID)
	attrs := []attribute.KeyValue{
		semconv.DriftModelIDKey.String(cfg.ModelID),
		semconv.DriftTypeKey.String(cfg.DriftType),
	}
	if cfg.DetectionMethod != "" {
		attrs = append(attrs, semconv.DriftDetectionMethodKey.String(cfg.DetectionMethod))
	}
	if cfg.ReferenceDataset != "" {
		attrs = append(attrs, semconv.DriftRefDatasetKey.String(cfg.ReferenceDataset))
	}
	if cfg.ReferencePeriod != "" {
		attrs = append(attrs, semconv.DriftRefPeriodKey.String(cfg.ReferencePeriod))
	}
	if cfg.Threshold != nil {
		attrs = append(attrs, semconv.DriftThresholdKey.Float64(*cfg.Threshold))
	}

	ctx, span := d.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &DriftDetection{span: span}
}

// SetScore records the drift score.
func (det *DriftDetection) SetScore(score float64) {
	det.span.SetAttributes(semconv.DriftScoreKey.Float64(score))
}

// SetResult records the detection result (e.g. "normal", "alert", "critical").
func (det *DriftDetection) SetResult(result string) {
	det.span.SetAttributes(semconv.DriftResultKey.String(result))
}

// SetMetrics records baseline vs current metric values.
func (det *DriftDetection) SetMetrics(baseline, current float64, metricName string) {
	det.span.SetAttributes(
		semconv.DriftBaselineMetricKey.Float64(baseline),
		semconv.DriftCurrentMetricKey.Float64(current),
		semconv.DriftMetricNameKey.String(metricName),
	)
}

// SetPValue records the statistical p-value.
func (det *DriftDetection) SetPValue(pValue float64) {
	det.span.SetAttributes(semconv.DriftPValueKey.Float64(pValue))
}

// SetSampleSize records the number of samples analyzed.
func (det *DriftDetection) SetSampleSize(size int) {
	det.span.SetAttributes(semconv.DriftSampleSizeKey.Int(size))
}

// SetAffectedSegments records the data segments affected by drift.
func (det *DriftDetection) SetAffectedSegments(segments []string) {
	det.span.SetAttributes(semconv.DriftAffectedSegmentsKey.StringSlice(segments))
}

// SetFeature records a feature name and its optional importance score.
func (det *DriftDetection) SetFeature(name string, importance *float64) {
	det.span.SetAttributes(semconv.DriftFeatureNameKey.String(name))
	if importance != nil {
		det.span.SetAttributes(semconv.DriftFeatureImportanceKey.Float64(*importance))
	}
}

// SetActionTriggered records the action triggered by drift detection.
func (det *DriftDetection) SetActionTriggered(action string) {
	det.span.SetAttributes(semconv.DriftActionTriggeredKey.String(action))
}

// End completes the drift detection span.
func (det *DriftDetection) End(err error) {
	if err != nil {
		det.span.SetStatus(codes.Error, err.Error())
		det.span.RecordError(err)
	} else {
		det.span.SetStatus(codes.Ok, "")
	}
	det.span.End()
}

// Span returns the underlying OTel span.
func (det *DriftDetection) Span() trace.Span { return det.span }

// ── Baseline ────────────────────────────────────────────────────────────────

// BaselineConfig holds configuration for a baseline management span.
type BaselineConfig struct {
	ModelID    string
	Operation  string // e.g. "create", "refresh"
	Dataset    string
	SampleSize *int
	Period     string
}

// DriftBaseline provides methods for enriching a drift baseline span.
type DriftBaseline struct {
	span trace.Span
}

// TraceBaseline starts a span for baseline establishment or refresh. The caller
// must call End() when done.
func (d *DriftDetectionInstrumentor) TraceBaseline(ctx context.Context, cfg BaselineConfig) (context.Context, *DriftBaseline) {
	if cfg.Operation == "" {
		cfg.Operation = "create"
	}

	spanName := fmt.Sprintf("drift.baseline %s %s", cfg.Operation, cfg.ModelID)
	attrs := []attribute.KeyValue{
		semconv.DriftModelIDKey.String(cfg.ModelID),
		semconv.DriftBaselineOperationKey.String(cfg.Operation),
	}
	if cfg.Dataset != "" {
		attrs = append(attrs, semconv.DriftBaselineDatasetKey.String(cfg.Dataset))
	}
	if cfg.SampleSize != nil {
		attrs = append(attrs, semconv.DriftBaselineSampleSizeKey.Int(*cfg.SampleSize))
	}
	if cfg.Period != "" {
		attrs = append(attrs, semconv.DriftBaselinePeriodKey.String(cfg.Period))
	}

	ctx, span := d.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &DriftBaseline{span: span}
}

// SetID records the baseline identifier.
func (b *DriftBaseline) SetID(baselineID string) {
	b.span.SetAttributes(semconv.DriftBaselineIDKey.String(baselineID))
}

// SetMetrics records baseline metrics as a JSON string.
func (b *DriftBaseline) SetMetrics(metricsJSON string) {
	b.span.SetAttributes(semconv.DriftBaselineMetricsKey.String(metricsJSON))
}

// SetFeatures records the feature names included in the baseline.
func (b *DriftBaseline) SetFeatures(features []string) {
	b.span.SetAttributes(semconv.DriftBaselineFeaturesKey.StringSlice(features))
}

// SetPreviousID records the identifier of the previous baseline.
func (b *DriftBaseline) SetPreviousID(previousID string) {
	b.span.SetAttributes(semconv.DriftBaselinePreviousIDKey.String(previousID))
}

// End completes the baseline span.
func (b *DriftBaseline) End(err error) {
	if err != nil {
		b.span.SetStatus(codes.Error, err.Error())
		b.span.RecordError(err)
	} else {
		b.span.SetStatus(codes.Ok, "")
	}
	b.span.End()
}

// Span returns the underlying OTel span.
func (b *DriftBaseline) Span() trace.Span { return b.span }

// ── Investigation ───────────────────────────────────────────────────────────

// InvestigationConfig holds configuration for a drift investigation span.
type InvestigationConfig struct {
	ModelID   string
	TriggerID string
}

// DriftInvestigation provides methods for enriching a drift investigation span.
type DriftInvestigation struct {
	span trace.Span
}

// TraceInvestigate starts a span for a drift investigation. The caller must call
// End() when done.
func (d *DriftDetectionInstrumentor) TraceInvestigate(ctx context.Context, cfg InvestigationConfig) (context.Context, *DriftInvestigation) {
	spanName := fmt.Sprintf("drift.investigate %s", cfg.ModelID)
	attrs := []attribute.KeyValue{
		semconv.DriftModelIDKey.String(cfg.ModelID),
		semconv.DriftInvestTriggerIDKey.String(cfg.TriggerID),
	}

	ctx, span := d.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &DriftInvestigation{span: span}
}

// SetRootCause records the root cause and its category.
func (inv *DriftInvestigation) SetRootCause(cause, category string) {
	inv.span.SetAttributes(
		semconv.DriftInvestRootCauseKey.String(cause),
		semconv.DriftInvestRootCauseCatKey.String(category),
	)
}

// SetImpact records the impact of the drift event.
func (inv *DriftInvestigation) SetImpact(affectedSegments []string, affectedUsersEstimate *int, blastRadius string) {
	inv.span.SetAttributes(semconv.DriftInvestAffectedSegmentsKey.StringSlice(affectedSegments))
	if affectedUsersEstimate != nil {
		inv.span.SetAttributes(semconv.DriftInvestAffectedUsersKey.Int(*affectedUsersEstimate))
	}
	if blastRadius != "" {
		inv.span.SetAttributes(semconv.DriftInvestBlastRadiusKey.String(blastRadius))
	}
}

// SetSeverity records the severity of the drift investigation.
func (inv *DriftInvestigation) SetSeverity(severity string) {
	inv.span.SetAttributes(semconv.DriftInvestSeverityKey.String(severity))
}

// SetRecommendation records the recommended action.
func (inv *DriftInvestigation) SetRecommendation(recommendation string) {
	inv.span.SetAttributes(semconv.DriftInvestRecommendationKey.String(recommendation))
}

// End completes the investigation span.
func (inv *DriftInvestigation) End(err error) {
	if err != nil {
		inv.span.SetStatus(codes.Error, err.Error())
		inv.span.RecordError(err)
	} else {
		inv.span.SetStatus(codes.Ok, "")
	}
	inv.span.End()
}

// Span returns the underlying OTel span.
func (inv *DriftInvestigation) Span() trace.Span { return inv.span }

// ── Remediation ─────────────────────────────────────────────────────────────

// RemediationConfig holds configuration for a drift remediation span.
type RemediationConfig struct {
	ModelID     string
	Action      string // e.g. "retrain", "rollback", "recalibrate"
	TriggerID   string
	Automated   bool
	InitiatedBy string
}

// DriftRemediation provides methods for enriching a drift remediation span.
type DriftRemediation struct {
	span trace.Span
}

// TraceRemediate starts a span for a drift remediation action. The caller must
// call End() when done.
func (d *DriftDetectionInstrumentor) TraceRemediate(ctx context.Context, cfg RemediationConfig) (context.Context, *DriftRemediation) {
	spanName := fmt.Sprintf("drift.remediate %s %s", cfg.Action, cfg.ModelID)
	attrs := []attribute.KeyValue{
		semconv.DriftModelIDKey.String(cfg.ModelID),
		semconv.DriftRemediationActionKey.String(cfg.Action),
		semconv.DriftRemediationAutomatedKey.Bool(cfg.Automated),
	}
	if cfg.TriggerID != "" {
		attrs = append(attrs, semconv.DriftRemediationTriggerIDKey.String(cfg.TriggerID))
	}
	if cfg.InitiatedBy != "" {
		attrs = append(attrs, semconv.DriftRemediationInitiatedByKey.String(cfg.InitiatedBy))
	}

	ctx, span := d.tracer.Start(ctx, spanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithAttributes(attrs...),
	)
	return ctx, &DriftRemediation{span: span}
}

// SetStatus records the remediation status.
func (rem *DriftRemediation) SetStatus(status string) {
	rem.span.SetAttributes(semconv.DriftRemediationStatusKey.String(status))
}

// SetRollbackTo records the model version to roll back to.
func (rem *DriftRemediation) SetRollbackTo(modelVersion string) {
	rem.span.SetAttributes(semconv.DriftRemediationRollbackToKey.String(modelVersion))
}

// SetRetrainDataset records the dataset used for retraining.
func (rem *DriftRemediation) SetRetrainDataset(dataset string) {
	rem.span.SetAttributes(semconv.DriftRemediationRetrainKey.String(dataset))
}

// SetValidationPassed records whether the remediation validation passed.
func (rem *DriftRemediation) SetValidationPassed(passed bool) {
	rem.span.SetAttributes(semconv.DriftRemediationValidPassedKey.Bool(passed))
}

// End completes the remediation span.
func (rem *DriftRemediation) End(err error) {
	if err != nil {
		rem.span.SetStatus(codes.Error, err.Error())
		rem.span.RecordError(err)
	} else {
		rem.span.SetStatus(codes.Ok, "")
	}
	rem.span.End()
}

// Span returns the underlying OTel span.
func (rem *DriftRemediation) Span() trace.Span { return rem.span }

File: telemetry/aitf/sdk/typescript/src/instrumentation/drift-detection.ts
/**
 * AITF Model Drift Detection Instrumentation.
 *
 * Provides structured tracing for model drift detection, baseline management,
 * drift investigation, and remediation. Aligned with CoSAI's identification of
 * model drift as a top-level AI incident category.
 */

import {
  trace,
  Span,
  SpanKind,
  SpanStatusCode,
  Tracer,
  TracerProvider,
} from "@opentelemetry/api";
import { DriftDetectionAttributes } from "../semantic-conventions/attributes";

const TRACER_NAME = "instrumentation.drift_detection";

// ---------------------------------------------------------------------------
// Option interfaces
// ---------------------------------------------------------------------------

/** Options for tracing a drift detection analysis. */
export interface TraceDetectOptions {
  modelId: string;
  driftType: string;
  detectionMethod?: string;
  referenceDataset?: string;
  referencePeriod?: string;
  threshold?: number;
}

/** Options for tracing baseline establishment or refresh. */
export interface TraceBaselineOptions {
  modelId: string;
  operation?: string;
  dataset?: string;
  sampleSize?: number;
  period?: string;
}

/** Options for tracing a drift investigation. */
export interface TraceInvestigateOptions {
  modelId: string;
  triggerId: string;
}

/** Options for tracing a drift remediation action. */
export interface TraceRemediateOptions {
  modelId: string;
  action: string;
  triggerId?: string;
  automated?: boolean;
  initiatedBy?: string;
}

// ---------------------------------------------------------------------------
// Helper classes
// ---------------------------------------------------------------------------

/** Helper for enriching a drift detection span. */
export class DriftDetection {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record the drift detection score. */
  setScore(score: number): void {
    this._span.setAttribute(DriftDetectionAttributes.SCORE, score);
  }

  /** Record the detection result (e.g. "alert", "normal", "warning"). */
  setResult(result: string): void {
    this._span.setAttribute(DriftDetectionAttributes.RESULT, result);
  }

  /** Record baseline vs. current metric comparison. */
  setMetrics(baseline: number, current: number, metricName: string): void {
    this._span.setAttribute(DriftDetectionAttributes.BASELINE_METRIC, baseline);
    this._span.setAttribute(DriftDetectionAttributes.CURRENT_METRIC, current);
    this._span.setAttribute(DriftDetectionAttributes.METRIC_NAME, metricName);
  }

  /** Record the statistical p-value from the detection test. */
  setPValue(pValue: number): void {
    this._span.setAttribute(DriftDetectionAttributes.P_VALUE, pValue);
  }

  /** Record the sample size used for detection. */
  setSampleSize(size: number): void {
    this._span.setAttribute(DriftDetectionAttributes.SAMPLE_SIZE, size);
  }

  /** Record the segments affected by the drift. */
  setAffectedSegments(segments: string[]): void {
    this._span.setAttribute(DriftDetectionAttributes.AFFECTED_SEGMENTS, segments);
  }

  /** Record the feature that drifted and optionally its importance. */
  setFeature(name: string, importance?: number): void {
    this._span.setAttribute(DriftDetectionAttributes.FEATURE_NAME, name);
    if (importance !== undefined) {
      this._span.setAttribute(DriftDetectionAttributes.FEATURE_IMPORTANCE, importance);
    }
  }

  /** Record the action triggered in response to the drift. */
  setActionTriggered(action: string): void {
    this._span.setAttribute(DriftDetectionAttributes.ACTION_TRIGGERED, action);
  }
}

/** Helper for enriching a drift baseline span. */
export class DriftBaseline {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record the baseline identifier. */
  setId(baselineId: string): void {
    this._span.setAttribute(DriftDetectionAttributes.BASELINE_ID, baselineId);
  }

  /** Record the baseline metrics as a JSON string. */
  setMetrics(metricsJson: string): void {
    this._span.setAttribute(DriftDetectionAttributes.BASELINE_METRICS, metricsJson);
  }

  /** Record the features captured in the baseline. */
  setFeatures(features: string[]): void {
    this._span.setAttribute(DriftDetectionAttributes.BASELINE_FEATURES, features);
  }

  /** Record the previous baseline identifier that this replaces. */
  setPreviousId(previousId: string): void {
    this._span.setAttribute(DriftDetectionAttributes.BASELINE_PREVIOUS_ID, previousId);
  }
}

/** Helper for enriching a drift investigation span. */
export class DriftInvestigation {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record the identified root cause and its category. */
  setRootCause(cause: string, category: string): void {
    this._span.setAttribute(DriftDetectionAttributes.INVESTIGATION_ROOT_CAUSE, cause);
    this._span.setAttribute(DriftDetectionAttributes.INVESTIGATION_ROOT_CAUSE_CATEGORY, category);
  }

  /** Record the impact assessment of the drift. */
  setImpact(options: {
    affectedSegments: string[];
    affectedUsersEstimate?: number;
    blastRadius?: string;
  }): void {
    this._span.setAttribute(
      DriftDetectionAttributes.INVESTIGATION_AFFECTED_SEGMENTS,
      options.affectedSegments,
    );
    if (options.affectedUsersEstimate !== undefined) {
      this._span.setAttribute(
        DriftDetectionAttributes.INVESTIGATION_AFFECTED_USERS,
        options.affectedUsersEstimate,
      );
    }
    if (options.blastRadius !== undefined) {
      this._span.setAttribute(
        DriftDetectionAttributes.INVESTIGATION_BLAST_RADIUS,
        options.blastRadius,
      );
    }
  }

  /** Record the severity of the investigation finding. */
  setSeverity(severity: string): void {
    this._span.setAttribute(DriftDetectionAttributes.INVESTIGATION_SEVERITY, severity);
  }

  /** Record the recommended remediation action. */
  setRecommendation(recommendation: string): void {
    this._span.setAttribute(DriftDetectionAttributes.INVESTIGATION_RECOMMENDATION, recommendation);
  }
}

/** Helper for enriching a drift remediation span. */
export class DriftRemediation {
  private readonly _span: Span;

  constructor(span: Span) {
    this._span = span;
  }

  /** Access the underlying OTel span. */
  get span(): Span {
    return this._span;
  }

  /** Record the remediation status. */
  setStatus(status: string): void {
    this._span.setAttribute(DriftDetectionAttributes.REMEDIATION_STATUS, status);
  }

  /** Record the model version to roll back to. */
  setRollbackTo(modelVersion: string): void {
    this._span.setAttribute(DriftDetectionAttributes.REMEDIATION_ROLLBACK_TO, modelVersion);
  }

  /** Record the dataset used for retraining. */
  setRetrainDataset(dataset: string): void {
    this._span.setAttribute(DriftDetectionAttributes.REMEDIATION_RETRAIN_DATASET, dataset);
  }

  /** Record whether the post-remediation validation passed. */
  setValidationPassed(passed: boolean): void {
    this._span.setAttribute(DriftDetectionAttributes.REMEDIATION_VALIDATION_PASSED, passed);
  }
}

// ---------------------------------------------------------------------------
// Instrumentor
// ---------------------------------------------------------------------------

/**
 * Instrumentor for model drift detection operations.
 *
 * Traces drift detection analyses, baseline management, investigations,
 * and remediation actions with AITF semantic convention attributes.
 */
export class DriftDetectionInstrumentor {
  private _tracerProvider: TracerProvider | null;
  private _tracer: Tracer | null = null;
  private _instrumented = false;

  constructor(tracerProvider?: TracerProvider) {
    this._tracerProvider = tracerProvider ?? null;
  }

  /** Enable drift detection instrumentation. */
  instrument(): void {
    const tp = this._tracerProvider ?? trace.getTracerProvider();
    this._tracer = tp.getTracer(TRACER_NAME);
    this._instrumented = true;
  }

  /** Disable drift detection instrumentation. */
  uninstrument(): void {
    this._tracer = null;
    this._instrumented = false;
  }

  /** Get the tracer instance, lazily initializing if needed. */
  getTracer(): Tracer {
    if (!this._tracer) {
      const tp = this._tracerProvider ?? trace.getTracerProvider();
      this._tracer = tp.getTracer(TRACER_NAME);
    }
    return this._tracer;
  }

  // -- Detection ------------------------------------------------------------

  /**
   * Trace a drift detection analysis.
   *
   * Returns a {@link DriftDetection} for manual span management, or accepts
   * a callback that receives the detection and automatically ends the span.
   */
  traceDetect(options: TraceDetectOptions): DriftDetection;
  traceDetect<T>(
    options: TraceDetectOptions,
    fn: (det: DriftDetection) => T,
  ): T;
  traceDetect<T>(
    options: TraceDetectOptions,
    fn?: (det: DriftDetection) => T,
  ): DriftDetection | T {
    const tracer = this.getTracer();

    const attributes: Record<string, string | number | boolean> = {
      [DriftDetectionAttributes.MODEL_ID]: options.modelId,
      [DriftDetectionAttributes.TYPE]: options.driftType,
    };
    if (options.detectionMethod !== undefined) {
      attributes[DriftDetectionAttributes.DETECTION_METHOD] = options.detectionMethod;
    }
    if (options.referenceDataset !== undefined) {
      attributes[DriftDetectionAttributes.REFERENCE_DATASET] = options.referenceDataset;
    }
    if (options.referencePeriod !== undefined) {
      attributes[DriftDetectionAttributes.REFERENCE_PERIOD] = options.referencePeriod;
    }
    if (options.threshold !== undefined) {
      attributes[DriftDetectionAttributes.THRESHOLD] = options.threshold;
    }

    const spanName = `drift.detect ${options.driftType} ${options.modelId}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const det = new DriftDetection(otelSpan);
          try {
            const result = fn(det);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new DriftDetection(otelSpan);
  }

  // -- Baseline -------------------------------------------------------------

  /**
   * Trace baseline establishment or refresh.
   *
   * Returns a {@link DriftBaseline} for manual span management, or accepts
   * a callback that receives the baseline and automatically ends the span.
   */
  traceBaseline(options: TraceBaselineOptions): DriftBaseline;
  traceBaseline<T>(
    options: TraceBaselineOptions,
    fn: (baseline: DriftBaseline) => T,
  ): T;
  traceBaseline<T>(
    options: TraceBaselineOptions,
    fn?: (baseline: DriftBaseline) => T,
  ): DriftBaseline | T {
    const tracer = this.getTracer();
    const operation = options.operation ?? "create";

    const attributes: Record<string, string | number | boolean> = {
      [DriftDetectionAttributes.MODEL_ID]: options.modelId,
      [DriftDetectionAttributes.BASELINE_OPERATION]: operation,
    };
    if (options.dataset !== undefined) {
      attributes[DriftDetectionAttributes.BASELINE_DATASET] = options.dataset;
    }
    if (options.sampleSize !== undefined) {
      attributes[DriftDetectionAttributes.BASELINE_SAMPLE_SIZE] = options.sampleSize;
    }
    if (options.period !== undefined) {
      attributes[DriftDetectionAttributes.BASELINE_PERIOD] = options.period;
    }

    const spanName = `drift.baseline ${operation} ${options.modelId}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const baseline = new DriftBaseline(otelSpan);
          try {
            const result = fn(baseline);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new DriftBaseline(otelSpan);
  }

  // -- Investigation --------------------------------------------------------

  /**
   * Trace a drift investigation.
   *
   * Returns a {@link DriftInvestigation} for manual span management, or
   * accepts a callback that receives the investigation and automatically
   * ends the span.
   */
  traceInvestigate(options: TraceInvestigateOptions): DriftInvestigation;
  traceInvestigate<T>(
    options: TraceInvestigateOptions,
    fn: (inv: DriftInvestigation) => T,
  ): T;
  traceInvestigate<T>(
    options: TraceInvestigateOptions,
    fn?: (inv: DriftInvestigation) => T,
  ): DriftInvestigation | T {
    const tracer = this.getTracer();

    const attributes: Record<string, string | number | boolean> = {
      [DriftDetectionAttributes.MODEL_ID]: options.modelId,
      [DriftDetectionAttributes.INVESTIGATION_TRIGGER_ID]: options.triggerId,
    };

    const spanName = `drift.investigate ${options.modelId}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const inv = new DriftInvestigation(otelSpan);
          try {
            const result = fn(inv);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new DriftInvestigation(otelSpan);
  }

  // -- Remediation ----------------------------------------------------------

  /**
   * Trace a drift remediation action.
   *
   * Returns a {@link DriftRemediation} for manual span management, or
   * accepts a callback that receives the remediation and automatically
   * ends the span.
   */
  traceRemediate(options: TraceRemediateOptions): DriftRemediation;
  traceRemediate<T>(
    options: TraceRemediateOptions,
    fn: (rem: DriftRemediation) => T,
  ): T;
  traceRemediate<T>(
    options: TraceRemediateOptions,
    fn?: (rem: DriftRemediation) => T,
  ): DriftRemediation | T {
    const tracer = this.getTracer();
    const automated = options.automated ?? false;

    const attributes: Record<string, string | number | boolean> = {
      [DriftDetectionAttributes.MODEL_ID]: options.modelId,
      [DriftDetectionAttributes.REMEDIATION_ACTION]: options.action,
      [DriftDetectionAttributes.REMEDIATION_AUTOMATED]: automated,
    };
    if (options.triggerId !== undefined) {
      attributes[DriftDetectionAttributes.REMEDIATION_TRIGGER_ID] = options.triggerId;
    }
    if (options.initiatedBy !== undefined) {
      attributes[DriftDetectionAttributes.REMEDIATION_INITIATED_BY] = options.initiatedBy;
    }

    const spanName = `drift.remediate ${options.action} ${options.modelId}`;

    if (fn) {
      return tracer.startActiveSpan(
        spanName,
        { kind: SpanKind.INTERNAL, attributes },
        (otelSpan) => {
          const rem = new DriftRemediation(otelSpan);
          try {
            const result = fn(rem);
            if (result instanceof Promise) {
              return result
                .then((val) => {
                  otelSpan.setStatus({ code: SpanStatusCode.OK });
                  otelSpan.end();
                  return val;
                })
                .catch((err) => {
                  otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
                  otelSpan.recordException(err);
                  otelSpan.end();
                  throw err;
                }) as T;
            }
            otelSpan.setStatus({ code: SpanStatusCode.OK });
            otelSpan.end();
            return result;
          } catch (err) {
            otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: String(err) });
            otelSpan.recordException(err as Error);
            otelSpan.end();
            throw err;
          }
        },
      );
    }

    const otelSpan = tracer.startSpan(spanName, {
      kind: SpanKind.INTERNAL,
      attributes,
    });
    return new DriftRemediation(otelSpan);
  }
}
